File: app/src/test/java/org/mifos/mobile/UserDetailsPresenterTest.java
package org.mifos.mobile;

import android.content.Context;

import org.junit.After;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mifos.mobile.api.DataManager;
import org.mifos.mobile.api.local.PreferencesHelper;
import org.mifos.mobile.models.client.Client;
import org.mifos.mobile.models.notification.NotificationRegisterPayload;
import org.mifos.mobile.presenters.UserDetailsPresenter;
import org.mifos.mobile.ui.views.UserDetailsView;
import org.mifos.mobile.util.RxSchedulersOverrideRule;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;

import io.reactivex.Observable;
import okhttp3.ResponseBody;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@RunWith(MockitoJUnitRunner.class)
public class UserDetailsPresenterTest {

    @Rule
    public final RxSchedulersOverrideRule mOverrideSchedulersRule = new RxSchedulersOverrideRule();

    @Mock
    Context context;

    @Mock
    DataManager dataManager;

    @Mock
    UserDetailsView view;

    @Mock
    PreferencesHelper preferencesHelper;

    @Mock
    ResponseBody responseBody;

    private UserDetailsPresenter presenter;
    private Client client;

    @Before
    public void setUp() throws Exception {
        presenter = new UserDetailsPresenter(context, dataManager, preferencesHelper);
        client = FakeRemoteDataSource.getCurrentClient();
        presenter.attachView(view);
    }

    @After
    public void tearDown() throws Exception {
        presenter.detachView();
    }

    @Test
    public void testGetUserDetails() {
        when(dataManager.getCurrentClient())
                .thenReturn(Observable.just(client));
        presenter.getUserDetails();
        verify(view).showUserDetails(client);
    }

    @Test
    public void testGetUserDetailsFails() {
        when(dataManager.getCurrentClient())
                .thenReturn(Observable.<Client>error(new Throwable()));
        presenter.getUserDetails();

        verify(view).showError(context
                .getString(R.string.error_client_not_found));
    }

    @Test
    public void testRegisterNotification() {
        String token = "1";
        when(dataManager.registerNotification(any(NotificationRegisterPayload.class)))
                .thenReturn(Observable.just(responseBody));
        presenter.registerNotification(token);

        verify(preferencesHelper).setSentTokenToServer(true);
        verify(preferencesHelper).saveGcmToken(token);
    }
}

File: app/src/test/java/org/mifos/mobile/UserDetailsPresenterTest.kt
package org.mifos.mobile

import android.content.Context

import io.reactivex.Observable

import okhttp3.ResponseBody

import org.junit.After
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith

import org.mifos.mobile.api.DataManager
import org.mifos.mobile.api.local.PreferencesHelper
import org.mifos.mobile.models.client.Client
import org.mifos.mobile.models.notification.NotificationRegisterPayload
import org.mifos.mobile.presenters.UserDetailsPresenter
import org.mifos.mobile.ui.views.UserDetailsView
import org.mifos.mobile.util.RxSchedulersOverrideRule

import org.mockito.ArgumentMatchers
import org.mockito.Mock
import org.mockito.Mockito
import org.mockito.junit.MockitoJUnitRunner

@RunWith(MockitoJUnitRunner::class)
class UserDetailsPresenterTest {
    @Rule
    @JvmField
    val mOverrideSchedulersRule = RxSchedulersOverrideRule()

    @Mock
    var context: Context? = null

    @Mock
    var dataManager: DataManager? = null

    @Mock
    var view: UserDetailsView? = null

    @Mock
    var preferencesHelper: PreferencesHelper? = null

    @Mock
    var responseBody: ResponseBody? = null
    private var presenter: UserDetailsPresenter? = null
    private var client: Client? = null

    @Before
    @Throws(Exception::class)
    fun setUp() {
        presenter = UserDetailsPresenter(context!!, dataManager!!, preferencesHelper!!)
        client = FakeRemoteDataSource.currentClient
        presenter?.attachView(view)
    }

    @After
    @Throws(Exception::class)
    fun tearDown() {
        presenter?.detachView()
    }

    @Test
    fun testGetUserDetails() {
        Mockito.`when`(dataManager?.currentClient)
                .thenReturn(Observable.just(client))
        presenter?.userDetails
        Mockito.verify(view)?.showUserDetails(client)
    }

    @Test
    fun testGetUserDetailsFails() {
        Mockito.`when`(dataManager?.currentClient)
                .thenReturn(Observable.error(Throwable()))
        presenter?.userDetails
        Mockito.verify(view)?.showError(context
                ?.getString(R.string.error_client_not_found))
    }

    @Test
    fun testRegisterNotification() {
        val token = "1"
        Mockito.`when`<Observable<ResponseBody?>?>(dataManager?.registerNotification(ArgumentMatchers.any(NotificationRegisterPayload::class.java)))
                .thenReturn(Observable.just(responseBody))
        presenter?.registerNotification(token)
        Mockito.verify(preferencesHelper)?.setSentTokenToServer(true)
        Mockito.verify(preferencesHelper)?.saveGcmToken(token)
    }
}